src/nxt_otel.c

```diff
@@ -0,0 +1,273 @@
+
+/*
+ * Copyright (C) F5, Inc.
+ */
+
+#include <math.h>
+
+#include <nxt_router.h>
+#include <nxt_http.h>
+#include <nxt_otel.h>
+#include <nxt_mp.h>
+#include <nxt_work_queue.h>
+#include <nxt_main.h>
+
+#define NXT_OTEL_TRACEPARENT_LEN 55
+#define NXT_OTEL_BODY_SIZE_TAG "body size"
+
+static inline void nxt_otel_trace_and_span_init(nxt_task_t *, nxt_http_request_t *);
+static inline void nxt_otel_span_collect(nxt_task_t *, nxt_http_request_t *);
+static void nxt_otel_span_add_headers(nxt_task_t *, nxt_http_request_t *);
+static void nxt_otel_span_add_body(nxt_http_request_t *);
+static void nxt_otel_error(nxt_task_t *, nxt_http_request_t *);
+
+inline void
+nxt_otel_test_and_call_state(nxt_task_t *t, nxt_http_request_t *r)
+{
+    // catches null state and unset flow status
+    if (!r->otel || !r->otel->status) {
+        return;
+    }
+
+    switch (r->otel->status) {
+    case NXT_OTEL_INIT_STATE:
+        nxt_otel_trace_and_span_init(t, r);
+        break;
+    case NXT_OTEL_HEADER_STATE:
+        nxt_otel_span_add_headers(t, r);
+        break;
+    case NXT_OTEL_BODY_STATE:
+        nxt_otel_span_add_body(r);
+        break;
+    case NXT_OTEL_COLLECT_STATE:
+        nxt_otel_span_collect(t, r);
+        break;
+    case NXT_OTEL_ERROR_STATE:
+        nxt_otel_error(t, r);
+        break;
+    }
+}
+
+static inline void
+nxt_otel_state_transition(nxt_otel_state_t *state, nxt_otel_status_t status)
+{
+    if (status == NXT_OTEL_ERROR_STATE || state->status != NXT_OTEL_ERROR_STATE) {
+        state->status = status;
+    }
+}
+
+static inline void
+nxt_otel_trace_and_span_init(nxt_task_t *t, nxt_http_request_t *r)
+{
+    r->otel->trace =
+        nxt_otel_get_or_create_trace(r->otel->trace_id);
+    if (!r->otel->trace)
+    {
+        nxt_log(t, NXT_LOG_ERR, "error generating otel span");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_HEADER_STATE);
+}
+
+static void
+nxt_otel_span_add_headers(nxt_task_t *t, nxt_http_request_t *r)
+{
+    nxt_http_field_t *f, *cur;
+    u_char *traceval, *name_cur, *val_cur;
```

Comment:
I could do that, but I figured it would be more idiomatic to do all the variable declarations at the top like all the other unit functions. 

If we define it as part of the loop does it have to get pushed/popped from the stack with each iteration? Wouldnt it be more efficient to keep it here?

Reply:
In C anyway, nope.

Now we compile with `-std=gnu11`, one of things we can do is for loop initialisers. E.g.

```c
for (int i = 0; ...)
```

Which I will wholly encourage.

Reply:
Although in this case these variables are being heap allocated in the loop, which _is_ where you will take a performance hit.

It's sometimes better if there's some sensible upper bound size to just allocate a static buffer.